pkg/components/pluggable/discovery.go

```diff
@@ -0,0 +1,176 @@
+func withV1(serviceName string) string {
+	return fmt.Sprintf("%s.%s", protoPackage, serviceName)
+}
+
+type service struct {
+	// protoRef is the proto service name
+	protoRef string
+	// componentName is the component name that implements such service.
+	componentName string
+	// dialer is the used grpc connectiondialer.
+	dialer GRPCConnectionDialer
+}
+
+type reflectServiceClient interface {
+	ListServices() ([]string, error)
+}
+
+// serviceDiscovery returns all available discovered pluggable components services.
+// uses gRPC reflection package to list implemented services.
+func serviceDiscovery(reflectClientFactory func(string) (reflectServiceClient, *grpc.ClientConn, error)) ([]service, error) {
+	services := []service{}
+	componentsSocketPath := GetSocketFolderPath()
+	_, err := os.Stat(componentsSocketPath)
+
+	if os.IsNotExist(err) { // not exists is the same as empty.
+		return services, nil
+	}
+
+	log.Debugf("loading pluggable components under path %s", componentsSocketPath)
+	if err != nil {
+		return nil, err
+	}
+
+	files, err := os.ReadDir(componentsSocketPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not list pluggable components unix sockets")
+	}
+
+	for _, dirEntry := range files {
+		if dirEntry.IsDir() { // skip dirs
+			continue
+		}
+
+		f, err := dirEntry.Info()
+		if err != nil {
+			return nil, err
+		}
+
+		absPath := fmt.Sprintf("%s/%s", componentsSocketPath, f.Name())
+		if !utils.IsSocket(f) {
+			discoveryLog.Warnf("could not use socket for file %s", absPath)
+			continue
+		}
+
+		refctClient, conn, err := reflectClientFactory(absPath)
+		if err != nil {
+			return nil, err
+		}
+
+		serviceList, err := refctClient.ListServices()
+		if err != nil {
+			return nil, errors.Wrap(err, "unable to list services")
+		}
+
+		componentName := removeExt(f.Name())
+		dialer := func() (*grpc.ClientConn, error) {
+			return conn, nil
+		}
+		for _, svc := range serviceList {
+			services = append(services, service{
+				componentName: componentName,
+				protoRef:      svc,
+				dialer:        dialer,
+			})
+		}
+	}
+	log.Debugf("found %d pluggable component services", len(services)-1) // reflection api doesn't count.
+	return services, nil
+}
+
+// callback invoke callback function for each given service
+func callback(services []service) {
+	for _, service := range services {
+		callback, ok := onServiceDiscovered[service.protoRef]
+		if !ok { // ignoring unknown service
+			continue
+		}
+		callback(service.componentName, service.dialer)
+		log.Infof("pluggable component '%s' was successfully registered for '%s'", service.componentName, service.protoRef)
+	}
+}
+
+// Discover discover the pluggable components and callback the service discovery with the given component name and grpc dialer.
+func Discover(ctx context.Context) error {
+	services, err := serviceDiscovery(func(socket string) (reflectServiceClient, *grpc.ClientConn, error) {
+		conn, err := SocketDial(socket, grpc.WithBlock())
```

Comment:
Should there be a timeout here?

Reply:
I think we can use the same context as we do for the entire discovery. I can choose an arbitrary timeout for each one individually but I guess it won't be sufficient because in case of a socket exists and a component is defined to use that socket, the runtime will panic in case of not being able to connect, so I think it's fair enough to use the parent context, what do you think?